src/table_of_contents_parser.py

```diff
@@ -0,0 +1,163 @@
+    None if not found.
+    """
+
+    def is_dotted_line(line: str) -> bool:
+        return len(line) >= 20 and line.count('.') >= len(line) * 0.1
+
+    lines = plain_text.split('\n')
+    dot_lines = [i for i in range(len(lines)) if is_dotted_line(lines[i])]
+
+    if len(dot_lines) == 0:
+        return None
+
+    dot_lines_med = dot_lines[len(dot_lines) // 2]
+
+    def is_outlier(line: int) -> bool:
+        return abs(line - dot_lines_med) > 1.5 * len(dot_lines)
+
+    dot_lines = [line for line in dot_lines if not is_outlier(line)]
+    margin = 30
+    start = max(0, dot_lines[0] - margin)
+    end = min(len(lines), dot_lines[-1] + margin)
+    return "\n".join(lines[start:end])
+
+
+def parse_toc_with_dots_multiple_columns(plain_text: str) \
+        -> List[Tuple[str, str, int]]:
+    lines = plain_text.split("\n")
+
+    matches = []
+    left_column = ""
+    right_column = ""
+
+    for line in lines:
+        match_iter = RE_TOC_WITH_DOTS.finditer(line)
+
+        try:
+            match = next(match_iter)
+            lookbehind, id, title, page = match.groups()
+            matches.append((lookbehind, id, title, page))
+            left_column += squash_whitespace(line[:match.start()]) + '\n'
+            right_column += squash_whitespace(line[match.end():]) + '\n'
+        except StopIteration:
+            # NOTE: This is the weak point. We have tried counting in also
+            # the surrounding page and section numbers.
+            spaces = list(re.finditer(r"\s{2,}", line))
+            if len(spaces) != 0:
+                largest_space = max(
+                    spaces, key=lambda match: match.end() - match.start())
+                left = line[:largest_space.start()]
+                right = line[largest_space.end():]
+                left_column += squash_whitespace(left) + '\n'
+                right_column += squash_whitespace(right) + '\n'
+
+    result = postprocess_matches(matches)
+    result += parse_toc_with_dots(left_column)
+    result += parse_toc_with_dots(right_column)
+    return result
+
+
+def parse_toc_with_dots(plain_text: str) -> List[Tuple[str, str, int]]:
+    matches = RE_TOC_WITH_DOTS.findall(plain_text, re.MULTILINE)
+    return postprocess_matches(matches) if matches is not None else []
+
+
+def parse_toc_without_dots(plain_text: str) -> List[Tuple[str, str, int]]:
+    matches = RE_TOC_WITHOUT_DOTS.findall(plain_text)
+    return postprocess_matches(matches) if matches is not None else []
+
+
+def postprocess_match(match: Tuple[str, str, str, str]) \
+        -> Optional[Tuple[str, str, int]]:
+    if match[0] is not None and len(match[0]) != 0:
+        return None
+
+    id, title, page_str = tuple(group.strip() for group in match[1:])
+    title = squash_whitespace(title)
+    title = re.sub(r"([A-Za-z])-\s", r"\1", title)
+    page = int(page_str)
+    return id, title, page
+
+
+def postprocess_matches(matches: List[Tuple[str, str, str, str]]) \
```

Comment:
If you know, that you might pass None as an argument, you can use Optional[str], which should be equal to Union[str, None]. It seems, that you check for None in the code itself, but this will increase code clarity and that is really useful in bigger python codebases.
Found using beartype.